<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0321ENSkillsNetwork928-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="Skills Network Logo"  />
    </a>
</p>


<h1> Review Of Accessing APIs</h1>

Estimated time needed: **30** minutes

## Objectives

After completing this lab, you will be able to:

*   Understand HTTP
*   Analyze HTTP Requests


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>
            <a href="#Overview-of-HTTP">Overview of HTTP </a>
            <ul>
                <li><a href="#Uniform-Resource-Locator-(URL)">Uniform Resource Locator: URL</a></li>
                 <li><a href="#Request">Request</a></li>
                <li><a href="#Response">Response</a></li>
            </ul>
        </li>
        <li>
            <a href="#Requests-in-Python">Requests in Python  </a>
            <ul>
                <li><a href="#Get-Request-with-URL-Parameters">Get Request with URL Parameters</a></li>
                <li><a href="#Post-Requests">Post Requests </a></li>

</ul>

</div>

<hr>


## Overview of HTTP


When you, the **client**, access a web page, your browser sends an **HTTP** request to the **server** where the page is hosted. The server tries to locate the desired **resource** by default "<code>index.html</code>". If your request is successful, the server will send the object to the client in an **HTTP response**, which includes information like the type of the **resource**, the length of the **resource**, and other information.

<p>
The figure below represents the process. The circle on the left represents the client, the circle on the right represents the web server. The table under the web server represents a list of resources stored in the web server. In  this case an <code>HTML</code> file, <code>png</code> image, and <code>txt</code> file .
</p>
<p>
The <b>HTTP</b> protocol enables you to send and receive information through the web including webpages, images, and other web resources. In this lab, you will explore the Requests library for interacting with the <code>HTTP</code> protocol. 
</p


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/reqest_basics.png" width="750" align="center">

</div>


## Uniform Resource Locator (URL)


Uniform resource locator (URL) is the common way to find resources on the web.  A URL can be broken down into three main parts:

<ul>
    <li><b>Scheme</b>: The protocol used, which in this lab will always be<code>http://</code>  </li>
    <li><b> Internet address or  Base URL</b>: Used to locate the resource. Examples include <code>www.ibm.com</code> and  <code> www.gitlab.com </code> </li>
    <li><b>Route</b>: Location on the web server for example: <code>/images/IDSNlogo.png</code> </li>
</ul>


You may also hear the term Uniform Resource Identifier (URI). URL are actually a subset of URIs. Another popular term is endpoint, which refers to the URL of an operation provided by a web server.


## Request


The process can be broken into the <b>Request</b> and <b>Response </b> process.  The request using the get method is partially illustrated below. In the start line we have the <code>GET</code> method, this is an <code>HTTP</code> method. Also the location of the resource  <code>/index.html</code> and the <code>HTTP</code> version. The Request header passes additional information with an <code>HTTP</code> request:


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/reqest_messege.png" width="400" align="center">
</div>


When an <code>HTTP</code> request is made, an <code>HTTP</code> method is sent, this tells the server what action to perform.  A list of several <code>HTTP</code> methods is shown below. We will review more examples later.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/SC6-zFofaLr-ILglB1xFGA/HTTP%20image.png" width="900" align="center">
</div>


## Response


The figure below represents the response; the response start line contains the version number <code>HTTP/1.0</code>, a status code (200) meaning success, followed by a descriptive phrase (OK). The response header contains useful information. Finally, we have the response body containing the requested file, an <code> HTML </code> document.  It should be noted that some requests have headers.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/kGeR47UyGjoLBL8R9yVH5Q/Response%20image.png" width="600" align="center">
</div> 


Some status code examples are shown in the table below, the prefix indicates the class. These are shown in yellow, with actual status codes shown in  white. Check out the following <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">link </a> for more descriptions.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/status_code.png" width="300" align="center">
</div>


### Install the required libraries


In [1]:
!pip install requests
!pip install pillow

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 131.6 MB/s eta 0:00:00


## Requests in Python


Requests is a Python Library that allows you to send <code>HTTP/1.1</code> requests easily. We can import the library as follows:


In [2]:
import requests

We will also use the following libraries:


In [3]:
import os 
from PIL import Image
from IPython.display import IFrame

You can make a <code>GET</code> request via the method <code>get</code> to [www.ibm.com](http://www.ibm.com/?utm_source=Exinfluencer&utm_content=000026UJ&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01&utm_medium=Exinfluencer&utm_term=10006555):


In [4]:

url='https://www.ibm.com/'
r=requests.get(url)


We have the response object <code>r</code>, this has information about the request, like the status of the request. You can view the status code using the attribute <code>status_code</code>.


In [5]:
r.status_code

200

You can view the request headers:


In [6]:
print(r.request.headers)

{'User-Agent': 'python-requests/2.32.3', 'Accept-Encoding': 'gzip, deflate, br, zstd', 'Accept': '*/*', 'Connection': 'keep-alive', 'Cookie': '_abck=6778ECE23BD6B98D23448AEB37D528B6~-1~YAAQqYnMFxWonayZAQAAv0ycyw7rLVV0FzCDMZONu+VIyNUDJ/K2aBD0UKmmGkZ1YCGle2KrMbiODKlN40l6HgHnCvS6fPYmNZzIKrXLjPNqyUaE7Cviy9pI+gH7OYUYm2J2UaGJGo54IRx9PSA+UQC0AdxxnSFh3mN4xyKdw79h/T7XiBuq82UFDRNf2pwVH4PxW4Y4XhM1qU1Jl4O5xJslu6xVbNQGdxrUq6ODUOTnTHD6/QnhGhgNdB0Q2QTNuPXgAgwL/81PFu2tGBPDnQDkrKLDEn2i8/oncYeGxwLzRwEQLLxMGBxvC/qBgvWB9fTZjJ8nG4tPg5HpIk6XdGTEX5rL9oGgWghei69k/HCbs1mz7/1MjXU27RvcDH9l3ghWupR6MWkfC2elTvHUMmQZ3V6de9FZuyyzYnlvRVnzMJfLKXBo08sRKXCQnzORbaQ=~-1~-1~-1~-1~-1; bm_sz=59A88D1462813E08A7816205D0ECF7AF~YAAQqYnMFxaonayZAQAAv0ycyx3zCJ6c3xc0OZZ/BgWPdYIttQVjua4XhccrqROmRo8XaF0f0Piwj5a/6Kr3NEtplvwajqQdco4laFUOolWXVgkvyQBeD1jlr5S4O6F2dpqbP3LgD1xess32UiqCcL4o4Kpeya/W1MYdJ6elooWIygzQDrU+pIIT55YD8FJ5tRunO5wWF54OyK3Egx8HsifgxEzQi8jsx1JkEjRlZr9NX/qJDN1YWxMmuk7mZP0E/iJ4C1bqQmjkVakhCqklqJvRl4hWLvZZkkZrtfHXZE5N1gqNchr

You can view the request body, in the following line, as there is nobody for a get request we get <code>None</code>:


In [7]:
print("request body:", r.request.body)

request body: None


You can view the <code>HTTP</code> response header using the attribute <code>headers</code>. This returns a python dictionary of <code>HTTP</code> response headers.


In [8]:
header=r.headers
print(r.headers)

{'Content-Security-Policy': 'upgrade-insecure-requests', 'x-frame-options': 'SAMEORIGIN', 'Last-Modified': 'Thu, 09 Oct 2025 23:16:23 GMT', 'ETag': '"2d188-640c1fd5c624a-gzip"', 'Accept-Ranges': 'bytes', 'Content-Type': 'text/html;charset=utf-8', 'X-Content-Type-Options': 'nosniff', 'Cache-Control': 'max-age=600', 'Expires': 'Fri, 10 Oct 2025 01:04:02 GMT', 'X-Akamai-Transformed': '0 - 0 -', 'Content-Encoding': 'gzip', 'Date': 'Fri, 10 Oct 2025 00:54:02 GMT', 'Content-Length': '36729', 'Connection': 'keep-alive', 'Vary': 'Accept-Encoding', 'Strict-Transport-Security': 'max-age=31536000'}


You can obtain the date the request was sent using the key <code>Date</code>.


In [ ]:
header['date']

<code>Content-Type</code> indicates the type of data:


In [ ]:
header['Content-Type']

You can also check the <code>encoding</code>:


In [ ]:
 r.encoding

As the <code>Content-Type</code> is <code>text/html</code> you can use the attribute <code>text</code> to display the <code>HTML</code> in the body. You can review the first 100 characters:


In [ ]:
r.text[0:100]

You can load other types of data for non-text requests, like images. Consider the URL of the following image:


In [ ]:
# Use single quotation marks for defining string
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/IDSNlogo.png'

You can make a get request:


In [ ]:
r=requests.get(url)

You can look at the response header:


In [ ]:
print(r.headers)

You can see the <code>'Content-Type'</code>


In [ ]:
r.headers['Content-Type']

An image is a response object that contains the image as a <a href="https://docs.python.org/3/glossary.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01#term-bytes-like-object">bytes-like object</a>. As a result, we must save it using a file object. First, you specify the <u>file path and
name</u>


In [ ]:
path=os.path.join(os.getcwd(),'image.png')

You save the file, in order to access the body of the response we use the attribute <code>content</code> then save it using the <code>open</code> function and write <code>method</code>:


In [ ]:
with open(path,'wb') as f:
    f.write(r.content)

You can view the image:


In [ ]:
Image.open(path)

<h3>Question: Download a file </h3>


Consider the following URL:


<code>URL = <https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt</code>


Write the commands to download the txt file in the given link.


In [ ]:
## Write your code here

<details><summary>Click here for the solution</summary>

```python
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt'
path=os.path.join(os.getcwd(),'example1.txt')
r=requests.get(url)
with open(path,'wb') as f:
    f.write(r.content)

```

</details>


## Get Request with URL Parameters


You can use the <b>GET</b> method to modify the results of your query, for example, retrieving data from an API. We send a <b>GET</b> request to the  server. As before, we have the <b>Base URL</b>, in the <b>Route</b> we append <code>/get</code>, this indicates we would like to preform a <code>GET</code> request. This is demonstrated in the following table:


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/HvW_UIwnCUtHR7SKIJHqtQ/base-image-file.jpg" width="400" align="center">
</div>


The Base URL is for <code>[http://httpbin.org/](http://httpbin.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01)</code>. It is a simple HTTP Request and Response service. The <code>URL</code> in Python is given by:


In [ ]:
url_get='http://httpbin.org/get'

A <a href="https://en.wikipedia.org/wiki/Query_string?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">query string</a> is a part of a uniform resource locator (URL), it sends other information to the web server. The start of the query is a <code>?</code>, followed by a series of parameter and value pairs, as shown in the table below. The first parameter name is <code>name</code> and the value is <code>Joseph</code>. The second parameter name is <code>ID</code> and the Value is <code>123</code>. Each pair, parameter, and value is separated by an equals sign, <code>=</code>.
The series of pairs is separated by the ampersand <code>&</code>.


<div class="alert alert-block alert-info" style="margin-top: 20px">
         <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/images/query_string.png" width="500" align="center">
</div>


To create a Query string, add a dictionary. The keys are the parameter names and the values are the value of the Query string.


In [ ]:
payload={"name":"Joseph","ID":"123"}

Then passing the dictionary <code>payload</code> to the <code>params</code> parameter of the <code> get()</code> function:


In [ ]:
r=requests.get(url_get,params=payload)

You can print out the <code>URL</code> and see the name and values.


In [ ]:
r.url

There is no request body.


In [ ]:
print("request body:", r.request.body)

You can print out the status code.


In [ ]:
print(r.status_code)

You can view the response as text:


In [ ]:
print(r.text)

You can look at the <code>'Content-Type'</code>.


In [ ]:
r.headers['Content-Type']

As the content <code>'Content-Type'</code> is in the <code>JSON</code> format, you can use the method <code>json()</code>, it returns a Python <code>dict</code>:


In [ ]:
r.json()

The key <code>args</code> has the name and values:


In [ ]:
r.json()['args']

## Post Requests


Like a <code>GET</code> request, a <code>POST</code> is used to send data to a server, but the <code>POST</code> request sends the data in a request body. In order to send the Post Request in Python, in the <code>URL</code> you can change the route to <code>POST</code>:


In [ ]:
url_post='http://httpbin.org/post'

This endpoint will expect data as a file or as a form. A form is convenient way to configure an HTTP request to send data to a server.


To make a <code>POST</code> request we use the <code>post()</code> function, the variable <code>payload</code> is passed to the parameter <code> data </code>:


In [ ]:
try:
    response = requests.post(url_post, data=payload)
    if response.status_code == 200:
        print("Response JSON:", response.json()) 
    else:
        print(f"HTTP Error: {response.status_code} - {response.reason}")
except requests.exceptions.RequestException as e:
    print(f"An error occurred: {e}")

Comparing the URL from the response object of the <code>GET</code> and <code>POST</code> request you see the <code>POST</code> request has no name or value pairs.


In [ ]:
print("POST request URL:",response.url )
print("GET request URL:",r.url)

You can compare the <code>POST</code> and <code>GET</code> request body, you see only the <code>POST</code> request has a body:


In [ ]:
print("POST request body:",response.request.body)
print("GET request body:",r.request.body)

You can view the form as well:


In [ ]:
response.json()['form']

There is a lot more you can do. Check out <a href="https://requests.readthedocs.io/en/master/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Requests </a> for more.


<hr>

<p>Congratulations, you have completed your hands-on lab on Review Of Accessing APIs.
<hr>


## Authors

<p><a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01" target="_blank">Joseph Santarcangelo</a> 


### Other Contributors

<a href="https://www.linkedin.com/in/jiahui-mavis-zhou-a4537814a?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Mavis Zhou</a>


<!--
## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description           |
| ----------------- | ------- | ---------- | ---------------------------- |
| 2024-10-04        | 2.5     | Madhusudan | ID reviewed                  |
| 2024-05-23        | 2.5     | Madhusudan | ID reviewed                  |
| 2023-11-02        | 2.4     | Abhishek Gagneja | Updated instructions   |
| 2023-06-07        | 2.3     |Akansha Yadav| Spell Check                 |
| 2021-12-20        | 2.1     | Malika     | Updated the links            |
| 2020-09-02        | 2.0     | Simran     | Template updates to the file |
|                   |         |            |                              |
|                   |         |            |                              |

## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>
--!>


Copyright © IBM Corporation. All rights reserved.
